core/src/main/java/org/infinispan/cache/impl/CacheImpl.java

```diff
@@ -1499,6 +1409,49 @@ private Object executeCommandAndCommitIfNeeded(InvocationContext ctx, VisitableC
       return result;
    }
 
+   private <T> CompletableFuture<T> executeCommandAndCommitIfNeededAsync(InvocationContext ctx, VisitableCommand command) {
+      final boolean txInjected = isTxInjected(ctx);
+      CompletableFuture<T> cf;
+      final Transaction implicitTransaction;
+      try {
+         cf = (CompletableFuture<T>) invoker.invokeAsync(ctx, command);
+         if (txInjected) {
+            implicitTransaction = transactionManager.suspend();
```

Comment:
The interceptors must not try to access the TM transaction via the thread-local, only via the invocation context. So it's safe to suspend the TM transaction before calling `invokeAsync`.